### ISSUE TITLE
FormGroup.getRawValue() or patchValue() translates value undefined of a field to value null

### ISSUE THREAD

**Issue by zijianhuang** [angular/angular#40608]
# 🐞 bug report

### Affected Package
Angular 11.1  and 11.1.1 and likely previous releases

### Description

Each FormControl field of a FormGroup may be and should be initialized with undefined, like 
```js
	private createPhoneNumberForm(): FormGroup {
		return new FormGroup({
			fullNumber: new FormControl(undefined),
			id: new FormControl(undefined),
			entityId: new FormControl(undefined),
		}
		);
	}
```

if id or entityId is untouched, the value should remain undefined, rather than null. My backend respects and considers undefined, null and empty string of a JSON field. 

Currently after calling FormGroup.getRrawValue(), the raw data has the id field changed to null.
```
 raw: {"fullNumber":"0782828383","id":null,"entityId":null}
```

My workaround is to set the formControl value to undefined before calling getRawValue(), so the undefined field won't appear in the JSON payload. Rather clumsy. 

## 🔬 Minimal Reproduction

At https://stackblitz.com/angular/dqymemyngmr

Replace input-prefix-suffix-example.ts with this:
```js
import {Component, OnInit} from '@angular/core';
import { FormControl, FormBuilder, FormGroup, AbstractControl } from '@angular/forms';

/**
 * @title Inputs with prefixes and suffixes
 */
@Component({
  selector: 'input-prefix-suffix-example',
  templateUrl: 'input-prefix-suffix-example.html',
  styleUrls: ['input-prefix-suffix-example.css'],
})
export class InputPrefixSuffixExample  implements OnInit {
  model: PhoneNumber = {fullNumber: '0782828383'};

	phoneNumberForm: FormGroup;
  constructor(){
    this.phoneNumberForm=this.createPhoneNumberForm();
  }

	ngOnInit() {
		console.debug('phone model: ' + JSON.stringify(this.model));
		this.phoneNumberForm.patchValue(this.model);
		const raw = this.phoneNumberForm.getRawValue();
		console.debug('raw: '+ JSON.stringify(raw));
	}

	private createPhoneNumberForm(): FormGroup {
		return new FormGroup({
			fullNumber: new FormControl(undefined),
			id: new FormControl(undefined),
			entityId: new FormControl(undefined),
		}
		);
	}

  send(){
    const raw = this.phoneNumberForm.getRawValue();
    console.debug('raw: '+ JSON.stringify(raw));
    this.phoneNumberForm.controls['id'].setValue(undefined);
    const raw2 = this.phoneNumberForm.getRawValue();
    console.debug('raw2: '+ JSON.stringify(raw2));
  }
}

	export interface PhoneNumber {
		entityId?: string;
		fullNumber?: string;
		id?: number;
	}

```

And replace respective html with this:
```html
<form class="example-form" [formGroup]="phoneNumberForm">
  <mat-form-field class="example-full-width">
    <mat-label>Telephone</mat-label>
    <span matPrefix>+1 &nbsp;</span>
    <input type="tel" matInput formControlName="fullNumber">
    <mat-icon matSuffix (click)="send()">mode_edit</mat-icon>
  </mat-form-field>
</form>
```

You may check the browser console to see.

## 🔥 Exception or Error

This behavior causes the deserialization of my backend to fail, if I don't apply the workaround.

## 🌍  Your Environment

**Angular Version:**
```json
    "@angular/animations": "11.1.1",
    "@angular/cdk": "11.1.1",
    "@angular/common": "11.1.1",
    "@angular/compiler": "^11.1.1",
    "@angular/core": "^11.1.1",
    "@angular/flex-layout": "11.0.0-beta.33",
    "@angular/forms": "^11.1.1",
    "@angular/localize": "^11.1.1",
    "@angular/material": "11.1.1",
    "@angular/material-moment-adapter": "11.1.1",
    "@angular/platform-browser": "^11.1.1",
    "@angular/platform-browser-dynamic": "^11.1.1",
    "@angular/platform-server": "^11.1.1",
    "@angular/pwa": "0.1101.1",
    "@angular/router": "^11.1.1",
    "@angular/service-worker": "^11.1.1",

```

**Comment by RomanHubyak** [comment]
I use `new FormControl({value: undefined, disabled: true})` when I need to initialize a FormControl with undefined.

**Comment by zijianhuang** [comment]
@RomanHubyak

Thanks for the suggestion. However, it seems like just another workaround, so overall it is still looking clumsy in application codes.

Changing undefined to null is semantically incorrect.

**Comment by AndrewKushnir** [comment]
Hi @zijianhuang, thanks for creating this ticket.

I agree that it's not ideal that `undefined` can not be used as a default value provided at `FormControl` creation time. However changing this behavior would be relatively big breaking change. Also given [the `FormControl` constructor API](https://angular.io/api/forms/FormControl#constructor()), we need to support a case when developers want to provide validator or async validator (2nd or 3rd argument), but keep the value default (so `undefined` can be used as the first argument with the "use whatever default value is" meaning). So it seems like changing `FormControl` constructor API would also be needed to properly support the `undefined` case.

Given the fact that this update would require significant design work, introducing a breaking change and since there is a workaround available, I believe it would not be something that we'd be able to do in the nearest future. I'm closing this ticket for now, but I'm adding a link to the #13721 issue (that talks about improving form types) and we can revisit this as a part of that work.

Thank you.

**Comment by zijianhuang** [comment]
@AndrewKushnir 

Thank you very much for the explanation. And apparently this was a long standing defect not conforming to JavaScript semantic undefined, null and empty string. 

Since the fix may require significant design works, I would agree that the fix is merged with #13721 or alike which requires design works.

Look forward to the fix, hopefully in NG 12?

**Comment by AndrewKushnir** [comment]
Thanks for the reply @zijianhuang.

> Look forward to the fix, hopefully in NG 12?

v12 is in a few weeks from now, so if we decide to do that (including API changes to `FormControl`), it'd definitely be later, no ETA at this moment...
